### ISSUE TITLE
Problem with copy protection/teledisk images

### ISSUE THREAD

**Issue by Pudentane** [86Box/86Box#471]
Hi!

I've contacted you before about .PSI format. I understand the request is a little more complex than it seems, so I kind of gave up on it myself. However, there is one more issue which this time concerns your TD0 image support. I discovered it a while ago, just didn't report it at once. It appears that not all copy protection schemes are parsed correctly by 86box and certain games fail to run while asking for "KEY DISK" as though the protection tracks are not there. Yes, I've toggled "Turbo timings" on and off, tried different CPUs, but none of that seemed to help. Such games may be few, but they **run successfully in PCE**. It could be only one type of CP or more, I'm not sure. At this point I can only report Sid Meier's Pirates! and F-19 Stealth Fighter. According to some sources they use MPS protection 1 or 2, which introduce **overlapping sectors** on track 4 or 38-39 (side 0). Maybe that's the part your emulator can't quite swallow. It would be awesome if you could fix that or provide any workaround at least!

P. S. I've attached the games so you could test them out yourself:
[f19.zip](https://github.com/86Box/86Box/files/3898236/f19.zip)
[pirates.zip](https://github.com/86Box/86Box/files/3898441/pirates.zip)

**Comment by OBattler** [comment]
Yes, I don't think 86Box handles overlapping sector from TD0 at all, might be because I ported the code from MAME, and their code did not have overlaping sectors support. I'm going to look at PCE's code to see where they differ.

**Comment by OBattler** [comment]
How do I start the F-19 game? Either START.EXE or any of the ?GRAPHICS.EXE results in hang.

**Comment by OBattler** [comment]
So the F-19 game has some sectors with CRC error. One is huge (8192 bytes), so I had to add special handling for such sectors that are bigger than the entire track. Now it reads that sector, gets CRC error, then it reads another sector, gets CRC OK, then errors. The sector is loaded as OK. So I have no idea what it still doesn't like. That is, unless Disk 1 is not the key disk.

**Comment by OBattler** [comment]
Are we sure it's supposed to work from a TD0? Because at this point, it would be a better idea if I gave you my modded verison of the HxC program so that you could convert KryoFlux dumps of these floppies, if you have them, to my modified version of the HxC MFM format, which can then be mounted in 86Box. That would give you images with a proper FM or MFM encoded surface for every track, which is much better than handling such cases.

**Comment by OBattler** [comment]
OK, it seems the game expects to be able to write to one of the sectors, which will not work with Teledisk because 86Box does not support writing to Teledisk images as that would be a mess that would slow the emulator down, especially if advanced compression is used, but I did convert the inserted image into an 86F image (the emulator's native surface image format) using the emulator's "Export to 86F" function, and then mounted the 86F image, and then it did some more accesses, I need to see to what.

**Comment by Pudentane** [comment]
> OK, it seems the game expects to be able to write to one of the sectors

Yes, I had this very issue with some other games, they wouldn't start because the disk was "write-protected"! I think PCE handles it differently though, as it doesn't really modify a mounted image (timestamps remain unchanged)... I have no idea how Hampa pulled it off, but something tells me it creates a phantom dump from the original and temporarily substitutes the latter...

**Comment by OBattler** [comment]
OK, it turns out the F-19 game does not require writing. I got it to work from TD0 with the floppy drive in turbo mode now by adding data CRC error support to turbo mode. I think it checks for the CRC error'ed sectors to have specific data. To make this work correctly in non-turbo mode, I need to figure out, from looking at the dump of the huge sector, how the track layout is supposed to look like.

**Comment by OBattler** [comment]
Got this to work in non-turbo mode as well now, with a little cheating - when a byte in the sector's data field has been decoded from the floppy surface, if there's a turbo read handler present, it now ignores the decoded byte and instead gets a byte from the turbo read handler. That way I can return the exact sector data from the Teledisk image.

Please note that, however, this means such images are not exportable to 86F, or better, they are exportable but the game will not work with them. There is no way around that. This is where HxC MFM format mounting comes in handy, however, the original MFM format has a big problem in that it rounds the number of bitcells into bytes which means if a sector wraps around the track index in the middle of the ID or data field, there will be problems. My improved format fixes that but the author of HxC has so far ignored my e-mail where I sent him my code and asked him if it can be merged.

**Comment by Pudentane** [comment]
Thank you for all your efforts!

**Comment by OBattler** [comment]
The last commit fixes the F-19 game, but not the Pirates game - the Teledisk images of the Pirates game seem strange.

**Comment by OBattler** [comment]
And the latest commit now fixes Sid Meier's Pirates! as well. It turns out there was a better way of handling sectors with no ID field, and that TD0 images made at 300 kbps on a 360 rpm drive, which is the same as 250 kbps on a 300 rpm drive (so, standard double density), were being misidentified at seek time as 300 kbps on a 300 rpm drive, causing the sectors to be given a too big GAP3 size.

**Comment by Pudentane** [comment]
Excellent news!! I only hope whatever worked earlier will still work after these changes :)

Don't mean to bug you, but I've attached another batch of games not working in 86box, 2 of which don't work even in PCE - Jackal and Wizard and the Princess. I'm just really curious if it's Teledisk (after all, it's not good with all CP methods) or actually omission on part of Hampa. Something tells me the way PCE parses TD0 files isn't exactly perfect either, so I was wondering if you could improve it a little further :)

P. S. Gunship will probably work now as it is also from Microprose and should use MPS protection 1 or 2, but still :)

[protected.zip](https://github.com/86Box/86Box/files/3931306/protected.zip)

**Comment by Pudentane** [comment]
Sorry, one more thing. I've cleaned F19 dumps a bit, turns out they contained "phantom sectors" which is a pretty common issue that I now presume could be the result of dumping 300rpm disks in 360rpm drive (thanks to your insight!). I've attached the fixed dumps, tell me if there is any change in the game's behavior. This is very important to us!
[f19.zip](https://github.com/86Box/86Box/files/3934791/f19.zip)
